### PR TITLE
PATCH_METADATA:  Handle all fields specified by user

### PR DESCRIPTION
**Pull request by pawbu**
- using DeepDiff package to handle deep dictionary comparisons
- using regexp to handle lists of items

### CODE REVIEW

File: datastore_version_manager/domain/metadata_patcher.py [hunk 1]
```diff
@@ -1,14 +1,47 @@
...
+                         r"\['categoryTitle'\]",
+                         r"root\['measureVariables'\]\[\d+\]\['valueDomain'\]\['description'\]",
+                         r"root\['measureVariables'\]\[\d+\]\['valueDomain'\]\['measurementUnitDescription'\]",
+                         r"root\['attributeVariables'\]\[\d+\]\['name'\]",
+                         r"root\['attributeVariables'\]\[\d+\]\['description'\]",
+                         r"root\['attributeVariables'\]\[\d+\]\['valueDomain'\]\['description'\]"
+                     ])
+
+    if 'dictionary_item_added' in ddiff:
+        raise Exception(f"Is is not allowed to add new metadata fields: {ddiff['dictionary_item_added']}")
```
Comment (DanielElisenberg) [inline datastore_version_manager/domain/metadata_patcher.py:41]: Can we raise an explicit exception here? PatchMetadataError or similar

EDIT: and then catch it explicitly in the test below

File: datastore_version_manager/domain/metadata_patcher.py [hunk 2]
```diff
@@ -1,14 +1,47 @@
+from deepdiff import DeepDiff
```
Comment (DanielElisenberg) [inline datastore_version_manager/domain/metadata_patcher.py:1]: :100:

**Review by DanielElisenberg — COMMENTED** [review]
These will be sett to null when patching occurs in dataset builder transformer:
```
identifierVariables[].representedVariables[].validPeriod.start
identifierVariables[].representedVariables[].validPeriod.stop
measureVariable.representedVariables[].validPeriod.start
measureVariable.representedVariables[].validPeriod.stop
attributeVariables[].representedVariables[].validPeriod.start
attributeVariables[].representedVariables[].validPeriod.stop
```
Could this be represented in a test?

If we return the validated metadata from this function then these null variables will be saved to datastore

**Comment by pawbu** [comment]
I have now fixed the code to use output model of Sikt. DeepDiff did not have an easy way to patch specific fields with regard to validPeriods.

The field measureVariables[].valueDomain.sentinelAndMissingValues[].categoryTitle is mapped to measureVariable.representedVariables[].valueDomain.codeList[].category.

The user is aware that two fields in input model that could be relevant for PATCH_METADATA are not mapped to output model:
measureVariables.subjectFields.description
measureVariables.valueDomain.uriDefinition

File: datastore_version_manager/domain/metadata_patcher.py [hunk 3]
```diff
@@ -1,14 +1,81 @@
...
+    if 'values_changed' in ddiff:
+        raise PatchMetadataException(
+            f"There are changes in metadata fields that are not allowed: {ddiff['values_changed']}")
+
+    _patch_valid_periods(source_metadata, destination_metadata)
+
+    return source_metadata
+
+
+def _patch_valid_periods(source_metadata: dict, destination_metadata: dict):
```
Comment (DanielElisenberg) [inline datastore_version_manager/domain/metadata_patcher.py:56]: Will the toplevel object "temporalCoverage" miss patching?

File: tests/resources/patch_metadata/dataset_with_code_list/destination_SYNT_BEFOLKNING_KJOENN.json [hunk 4]
```diff
@@ -0,0 +1,113 @@
+{
+  "name": "SYNT_BEFOLKNING_KJOENN",
+  "populationDescription": "Kj\u00f8nn for en populasjon",
+  "languageCode": "no",
+  "temporality": "FIXED",
+  "subjectFields": [
+    "Helse",
+    "Helsetjenester"
+  ],
+  "temporalCoverage": {
```
Comment (DanielElisenberg) [inline tests/resources/patch_metadata/dataset_with_code_list/destination_SYNT_BEFOLKNING_KJOENN.json:10]: Add a whole number epoch date timestamp to destination and expected to make sure its patched

**Review by DanielElisenberg — COMMENTED** [review]
Some test resource comments

File: datastore_version_manager/domain/metadata_patcher.py (hunk 3 above)
Comment (pawbu) [reply datastore_version_manager/domain/metadata_patcher.py:56]: Good catch! I fix.

**Comment by pawbu** [comment]
This PR is replaced by https://github.com/statisticsnorway/microdata-validator/pull/19. Some tests will be moved into the other project.

**Pull request closed without merging**